add failing test for when parameters are deleted

gaucheph opened the pull request:
this adds a unit test (which fails currently) to assert that when
the watched test suite has a parameter removed, the BranchNode.merge
method also removes it from its tree.

gaucheph committed:
add failing test for when parameters are deleted

this adds a unit test (which fails currently) to assert that when
the watched test suite has a parameter removed, the BranchNode.merge
method also removes it from its tree.

gaucheph commented:
hey @ryanc414 i noticed this bug while tinkering with your project. i don't have a fix but i was hoping i could learn more about how this works by seeing what your fix would be. thanks and awesome project 👍

ryanc414 commented:
Thanks for raising, actually parametrization was not something I explicitly considered or tested with the recent changes I made to how tests are collected so it wouldn't surprise me if there are some bugs in how they are handled. I'll take a look at your test and see what improvements can be made in that area.

Commit:
Merge branch 'master' into remove-missing-leaves

Commit:
Fix up merge logic after re-collecting source file

Add a "merge_base" arg to the merge method. All nodes from the merge
base down are taken from the newer collection tree into the new one
directly instead of being merged recursively. This ensures that if
test functions, classes or parameters are removed from the source
file, they will be removed from our tree also.

ryanc414 commented:
Sorry I've not looked at this in ages, been focussing on other things!

Actually the problem you found is not specifically to do with parametrized testcases and more of a logical issue with how I was re-collecting and merging the collected nodes into the main tree after a file change event. Previously the merge was a literal merge, keeping old entries and adding or updating new ones. However, as you found this doesn't work as you would expect if any test function, class or template parameter has been removed. So now, after collecting a test file pytest_commander will take the new collection subtree from that file and insert it into the tree, replacing all old state related to that file.

Thanks again for spotting this! I'll release the fix soon.
